smt/expr.cpp

```diff
@@ -511,6 +512,59 @@ expr expr::lshr(const expr &rhs) const {
   return Z3_mk_bvlshr(ctx(), ast(), rhs());
 }
 
+expr expr::rotate(const expr &rhs, bool shift_left) const {
+  C(rhs);
+  if (isZero() || isAllOnes() || rhs.isZero())
+    return *this;
+
+  auto nbits = bits();
+  expr rhs_mod_width = rhs.urem(mkUInt(nbits, nbits));
+  if (rhs_mod_width.isZero())
+    return *this;
+
+  // FIXME: could do constant-folding, but *only* for i8/i16/i32/i64.
```

Comment:
Hmm, actually, i would love to hear some recommendations about this.
Currently the best i can think of is one of:
* changing `binop_[us]fold` to take `uint64_t (*native)(uint64_t, uint64_t, unsigned /*width*/)` callback.
* or, adding `binop_[us]fold_width` taking that param, and not touching the current `binop_[us]fold`.
* or, changing `binop_[us]fold` to be templated, and take lambda, not a pointer to callback.
(this doesn't cover the need for `ternaryop_[us]fold`..)

I'm not sure which way is most forward.